ee/codegen/src/project.ts

```diff
@@ -426,79 +404,84 @@ ${errors.slice(0, 3).map((err) => {
    */
   private getOrderedNodes(): WorkflowDataNode[] {
     const rawData = this.workflowVersionExecConfig.workflowRawData;
+    const nodes = rawData.nodes.filter(
+      (node): node is WorkflowDataNode => node.type !== "ENTRYPOINT"
+    );
 
     // Edge case: Workflow init only has two nodes of ENTRYPOINT and TERMINAL with no edge between them
     if (rawData.edges.length === 0) {
-      return rawData.nodes.filter(
-        (node): node is WorkflowDataNode => node.type !== "ENTRYPOINT"
-      );
+      return nodes;
     }
 
-    const nodesById = Object.fromEntries(
-      rawData.nodes.map((node) => [node.id, node])
-    );
+    const nodesById = Object.fromEntries(nodes.map((node) => [node.id, node]));
 
-    const edgesQueue = this.workflowContext.getEntrypointNodeEdges();
-    const edgesByPortId = this.workflowContext.getEdgesByPortId();
-    const processedEdges = new Set<WorkflowEdge>();
-    const processedNodeIds = new Set<string>();
+    // Create an adjacency list representation of the graph
+    const graph: Map<string, string[]> = new Map();
+    // Track in-degree (number of incoming edges) for each node
+    const inDegree: Map<string, number> = new Map();
 
-    const orderedNodes: WorkflowDataNode[] = [];
+    // Initialize graph with all nodes
+    nodes.forEach((node) => {
+      graph.set(node.id, []);
+      inDegree.set(node.id, 0);
+    });
 
-    while (edgesQueue.length > 0) {
-      const edge = edgesQueue.shift();
-      if (!edge) {
-        continue;
+    // Add edges to graph and count in-degrees
+    rawData.edges.forEach((edge) => {
+      const sourceId = edge.sourceNodeId;
+      const targetId = edge.targetNodeId;
```

Comment:
```suggestion
      const sourceNodeId = edge.sourceNodeId;
      const targetNodeId = edge.targetNodeId;
```

[nit] to disambiguate from the other ids that could be in play here (handles, edges, etc.) The algo here is a bit tricky, so will be a little more pedantic here trying to make it easier to follow